### ISSUE TITLE
return-undefined false positive with catch

### ISSUE THREAD

**Issue by seantanly** [palantir/tslint#3432]
### Bug Report

- __TSLint version__: 5.8.0
- __TypeScript version__: 2.6.1
- __Running TSLint via__: CLI

#### TypeScript code being linted

```ts
async function asyncWorker() {
  const state = Math.random()
  if (state > 0.8) { return "data" }
  if (state > 0.5) { throw new Error('not found') }
  throw new Error('unknown error')
}

function find() {
  return asyncWorker()
    .catch(err => {
      if (err.message === 'not found') {
        return undefined
      }
      throw err
    })
}
```

with `tslint.json` configuration:

```json
{
  "rules": {
    "return-undefined": { "severity": "error" }
  }
}
```

#### Actual behavior

```
`void` function should use `return;`, not `return undefined;`. (return-undefined)
  10 |     .catch(err => {
  11 |       if (err.message === 'not found') {
> 12 |         return undefined
     |        ^
  13 |       }
  14 |       throw err
  15 |     })
```

#### Expected behavior
No error

**Comment by ajafff** [comment]
This is expected behavior. Because the function passed to `.catch()` does not have a contextual type, its return type is inferred from the return statements in the body. Since there is only one `return undefined` the function has a return type of `undefined`. The rule correctly complains about `return undefined` where the return type is `undefined` or `void`.

**Comment by seantanly** [comment]
Thank you for the explanation. I suppose for such edge use case, suppressing it with 
`// tslint:disable-next-line:return-undefined` is the way to go.
